server/src/main/java/org/elasticsearch/action/bulk/BulkPrimaryExecutionContext.java

```diff
@@ -0,0 +1,321 @@
+     * end of the {@link BulkShardRequest#items()} array.
+     */
+    public boolean hasMoreOperationsToExecute() {
+        return currentIndex < request.items().length;
+    }
+
+
+    /** returns the name of the index the current request used */
+    public String getConcreteIndex() {
+        return getCurrentItem().index();
+    }
+
+    /** returns any primary response that was set by a previous primary */
+    public BulkItemResponse getPreviousPrimaryResponse() {
+        return getCurrentItem().getPrimaryResponse();
+    }
+
+    /** returns a translog location that is needed to be synced in order to persist all operations executed so far */
+    public Translog.Location getLocationToSync() {
+        assert hasMoreOperationsToExecute() == false;
+        return locationToSync;
+    }
+
+    private BulkItemRequest getCurrentItem() {
+        return request.items()[currentIndex];
+    }
+
+    /** returns the primary shard */
+    public IndexShard getPrimary() {
+        return primary;
+    }
+
+    /**
+     * sets the request that should actually be executed on the primary. This can be different then the request
+     * received from the user (specifically, an update request is translated to an indexing or delete request).
+     */
+    public void setRequestToExecute(DocWriteRequest writeRequest) {
+        assert currentItemState != ItemProcessingState.TRANSLATED &&
+            currentItemState != ItemProcessingState.EXECUTED &&
+            currentItemState != ItemProcessingState.COMPLETED :
+            currentItemState;
+        assert requestToExecute == null : requestToExecute;
+        requestToExecute = writeRequest;
+        currentItemState = ItemProcessingState.TRANSLATED;
+    }
+
+    /** returns the request that should be executed on the shard. */
+    public <T extends DocWriteRequest<T>> T getRequestToExecute() {
+        assert currentItemState == ItemProcessingState.TRANSLATED : currentItemState;
+        return (T) requestToExecute;
+    }
+
+    /** indicates that the current operation can not be completed and needs to wait for a new mapping from the master */
+    public void markAsRequiringMappingUpdate() {
+        assert currentItemState == ItemProcessingState.TRANSLATED: currentItemState;
+        currentItemState = ItemProcessingState.WAIT_FOR_MAPPING_UPDATE;
+        requestToExecute = null;
+        executionResult = null;
+    }
+
+    /** resets the current item state, prepare for a new execution */
+    public void resetForExecutionAfterRetry() {
+        assert currentItemState == ItemProcessingState.WAIT_FOR_MAPPING_UPDATE ||
+            currentItemState == ItemProcessingState.IMMEDIATE_RETRY: currentItemState;
+        currentItemState = ItemProcessingState.INITIAL;
+    }
+
+    /** completes the operation without doing anything on the primary */
+    public void markOperationAsNoOp(DocWriteResponse response) {
+        assert currentItemState != ItemProcessingState.EXECUTED &&
```

Comment:
I think this is only called in INITIAL state, so let's `assert currentItemState == INITIAL`